File: MyTOdo/app/src/main/java/com/example/mytodo/Adapter/TodoAdapter.java
package com.example.mytodo.Adapter;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mytodo.AddNewTask;
import com.example.mytodo.MainActivity;
import com.example.mytodo.Model.ToDoModel;
import com.example.mytodo.R;
import com.example.mytodo.Utils.DataBaseHandler;

public class TodoAdapter extends ListAdapter<ToDoModel, TodoAdapter.ViewHolder> {

    private MainActivity activity;
    private DataBaseHandler db;

    public TodoAdapter(DataBaseHandler db, MainActivity activity) {
        super(DIFF_CALLBACK);
        this.db = db;
        this.activity = activity;
    }

    private static final DiffUtil.ItemCallback<ToDoModel> DIFF_CALLBACK = new DiffUtil.ItemCallback<ToDoModel>() {
        @Override
        public boolean areItemsTheSame(@NonNull ToDoModel oldItem, @NonNull ToDoModel newItem) {
            // Compare unique IDs
            return oldItem.getId() == newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull ToDoModel oldItem, @NonNull ToDoModel newItem) {
            // Compare the content
            return oldItem.getTask().equals(newItem.getTask()) && oldItem.getStatus() == newItem.getStatus();
        }
    };

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.task_layout, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        db.openDatabase();
        ToDoModel item = getItem(position);
        holder.task.setText(item.getTask());
        holder.task.setChecked(toBoolean(item.getStatus()));

        holder.task.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                if (b) {
                    db.updateStatus(item.getId(), 1);
                } else {
                    db.updateStatus(item.getId(), 0);
                }
            }
        });
    }

    private boolean toBoolean(int n) {
        return n != 0;
    }

    public Context getContext() {
        return activity;
    }

    public void editItem(int position) {
        ToDoModel item = getItem(position);
        Bundle bundle = new Bundle();
        bundle.putInt("id", item.getId());
        bundle.putString("task", item.getTask());
        AddNewTask fragment = new AddNewTask();
        fragment.setArguments(bundle);
        fragment.show(activity.getSupportFragmentManager(), AddNewTask.TAG);
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        CheckBox task;

        ViewHolder(View view) {
            super(view);
            task = view.findViewById(R.id.todoCheckBox);
        }
    }
}

File: MyTOdo/app/src/main/java/com/example/mytodo/Adapter/TodoAdapter.kt
/*
package com.example.mytodo.Adapter

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.CompoundButton
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.mytodo.AddNewTask
import com.example.mytodo.MainActivity
import com.example.mytodo.Model.ToDoModel
import com.example.mytodo.R
import com.example.mytodo.Utils.DataBaseHandler

class TodoAdapter(private val db: DataBaseHandler, private val activity: MainActivity) :
    ListAdapter<ToDoModel, TodoAdapter.ViewHolder>(DIFF_CALLBACK) {

    companion object {
        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<ToDoModel>() {
            override fun areItemsTheSame(oldItem: ToDoModel, newItem: ToDoModel): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: ToDoModel, newItem: ToDoModel): Boolean {
                return oldItem.task == newItem.task && oldItem.status == newItem.status
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(R.layout.task_layout, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        db.openDatabase()
        val item = getItem(position)
        holder.task.text = item.task
        holder.task.isChecked = toBoolean(item.status)

        holder.task.setOnCheckedChangeListener { _: CompoundButton, b: Boolean ->
            val newStatus = if (b) 1 else 0
            db.updateStatus(item.id, newStatus)
        }
    }

    private fun toBoolean(n: Int): Boolean {
        return n != 0
    }

    val context: Context
        get() = activity

    fun deleteItem(position: Int) {
        val item = getItem(position)
        db.deleteTask(item.id)
        val currentList = currentList.toMutableList()
        currentList.removeAt(position)
        submitList(currentList)
    }

    fun editItem(position: Int) {
        val item = getItem(position)
        val bundle = Bundle()
        bundle.putInt("id", item.id)
        bundle.putString("task", item.task)
        val fragment = AddNewTask()
        fragment.arguments = bundle
        fragment.show(activity.supportFragmentManager, AddNewTask.TAG)
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val task: CheckBox = view.findViewById(R.id.todoCheckBox)
    }
}

*/
package com.example.mytodo.Adapter

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.CompoundButton
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.mytodo.AddNewTask
import com.example.mytodo.MainActivity
import com.example.mytodo.Model.ToDoModel
import com.example.mytodo.R
import com.example.mytodo.Utils.DataBaseHandler

class TodoAdapter(private val db: DataBaseHandler, private val activity: MainActivity) :
    ListAdapter<ToDoModel, TodoAdapter.ViewHolder>(DIFF_CALLBACK) {

    companion object {
        val DIFF_CALLBACK = object : DiffUtil.ItemCallback<ToDoModel>() {
            override fun areItemsTheSame(oldItem: ToDoModel, newItem: ToDoModel): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: ToDoModel, newItem: ToDoModel): Boolean {
                return oldItem.task == newItem.task && oldItem.status == newItem.status
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context)
            .inflate(R.layout.task_layout, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        db.openDatabase()
        val item = getItem(position)
        holder.task.text = item.task
        holder.task.isChecked = toBoolean(item.status)

        holder.task.setOnCheckedChangeListener { _: CompoundButton, b: Boolean ->
            val newStatus = if (b) 1 else 0
            db.updateStatus(item.id, newStatus)
        }

        // Add click listener for the whole item view
        holder.itemView.setOnClickListener {
            activity.navigateToDetail(item.task)
        }
    }

    private fun toBoolean(n: Int): Boolean {
        return n != 0
    }

    val context: Context
        get() = activity

    fun deleteItem(position: Int) {
        val item = getItem(position)
        db.deleteTask(item.id)
        val currentList = currentList.toMutableList()
        currentList.removeAt(position)
        submitList(currentList)
    }

    fun editItem(position: Int) {
        val item = getItem(position)
        val bundle = Bundle()
        bundle.putInt("id", item.id)
        bundle.putString("task", item.task)
        val fragment = AddNewTask()
        fragment.arguments = bundle
        fragment.show(activity.supportFragmentManager, AddNewTask.TAG)
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val task: CheckBox = view.findViewById(R.id.todoCheckBox)
    }
}